_posts/2019-02-05-configurable-builds-part-1.md

```diff
@@ -0,0 +1,280 @@
+that sets custom C++ compile options. `cc_binary(name = "mybinary", srcs =
+["mybinary.cc"], deps = [":mylibrary"], copts =
+["-DUSE_EXPERIMENTAL_FEATURES=1"])` might compile `mybinary.cc` with
+experimental features. But this won't happen for `:mylibrary` unless it sets
+its `copts` similarly. This makes attributes unsuitable for tasks like building
+a binary for a different CPU.
+
+If you imagine a build as a [graph](https://docs.bazel.build/versions/master/build-ref.html#dependencies)
+with parents above their dependencies, attributes change behavior
+[*up*](https://docs.bazel.build/versions/master/skylark/rules.html#providers)
+the graph while configuration changes behavior *down* the graph.
+
+## Platforms
+
+One of configurability's major efforts is designing a principled API for
+defining *platforms* and *toolchains*.
+
+While there are many kinds of "ways" you might want to build your project, in
+practice most developers want the flexibility to target different devices,
+OSes, CPUs, and other machine properties. The general term for this is
+*platforms*. Being able to reason about how your code interacts with platforms
+is the basis of [multi-platform software](https://en.wikipedia.org/wiki/Cross-platform_software).
+
+*Toolchains* are the set of programs that build your code. A C++ toolchain
+includes a compiler and linker and the flags that invoke them "correctly".
+Different platforms require different toolchains. gcc might be the compiler of
+choice for Linux while Xcode is used on the Mac. Even the same platform might
+use different toolchains. For example, maybe you want to use an experimental
+compiler.
+
+Not only are these concepts cross-language, but it's *important* that languages
+treat them consistently. Otherwise you get an uncoordinated mess of
+language-specific code that makes it hard to combine different languages in a
+single project.
+
+One of the reasons
+
+    $ bazel build //my:binary --cpu=arm --crosstool_top=//my:custom_toolchain --define MYFEATURE=1
+
+doesn't work well is that `--cpu` and `--crosstool_top` are C++-specific. What
+if you want to build Java libraries with C++ dependencies for both Linux and
+Android? This impacts not just C++, but also the JDK and maybe even which
+support libraries you need.
+
+Not only are `--cpu` and `-crosstool_top` insufficiently expressive, but they
+might actively sabotoge you. Consider this use of
+[select](https://docs.bazel.build/versions/master/configurable-attributes.html):
+
+```python
+java_library(
+    name = "supporting_library",
+	deps = select({
+	    ":android": [":extra_required_android_deps"],
+		"//conditions:default": []
+    }))
+```
+
+If you use `--cpu` and `crosstool_top` to define what `:android` means, what
+happens when an app supports a new Android phone with a different CPU?
+`:android` won't trigger, the app won't get required Android support libraries,
+and it will crash.
+
+### A Better Way
```

Comment:
It takes a while to get to this point, which is the focus of the article. I think that many people will see a long "History" section and may stop reading before they get to this section. Is it possible to make the History section more concise, rename it to something related to the problem (e.g., "Challenges with multi-platform builds"), and then get to the solution a little faster?

Reply:
Taking into account your other suggestion, I moved the bottom of the "History" section to the top as a lead, IMO to the benefit of both sections. :)

It shortens this section, which I agree is desirable. But it doesn't shorten it much. And I'm hesitant to cull things much more.

A key goal for me is that I'm not just writing about multiplatform builds. I'm trying to tell configurability's larger story and show how the different pieces fit together. So this broader context isn't just nice to have but in my view core content. The overall structure I'm imagining across both posts is to frame the story, go into platforms as one major piece of it, go into other major pieces, then tie it all up at the end. So there is a flow there I don't want to disrupt too much.

Another way to get to the cool new platform stuff earlier would be to move the "Configuration vs. Attributes" section to the end. That's admittedly a bit of a side point. But since it's a side point to the broader story, and not specifically platforms, I feel it'd be even more disjoint to push it to the end.